Add timeline plot with matplotlib as backend

eukaryo opened the pull request:
<!-- Thank you for creating a pull request! In general, we merge your pull request after it gets two or more approvals. To proceed to the review process by the maintainers, please make sure that the PR meets the following conditions: (1) it passes all CI checks, and (2) it is neither in the draft nor WIP state. If you wish to discuss the PR in the draft state or need any other help, please mention the Optuna development team in the PR. -->

## Motivation
<!-- Describe your motivation why you will submit this PR. This is useful for reviewers to understand the context of PR. -->

cf: https://github.com/optuna/optuna/pull/4470

In addition to plotly, I want to implement timeline plot using matplotlib as backend.

## Description of the changes
<!-- Describe the changes in this PR. -->

- Add timeline plot using matplotlib in optuna/visualization/matplotlib/_timeline.py.
- Add Unit tests in tests/visualization_tests/matplotlib_tests/test_timeline.py.

![test-fig-matplotlib](https://user-images.githubusercontent.com/22315555/227133045-b471f2aa-d40d-412a-bfda-73c8994e5fef.png)

HideakiImamura commented on optuna/visualization/matplotlib/_timeline.py:
```diff
@@ -0,0 +1,119 @@
+    for state, color in _cm.items():
+        if len([b for b in info.bars if b.state == state]) > 0:
+            legend_handles.append(patches.Patch(color=color, label=state.name))
+    ax.legend(handles=legend_handles, loc="lower right")
```
How about making the legend outside the plot using the `bbox_to_anchor` argument?

See https://stackoverflow.com/questions/4700614/how-to-put-the-legend-outside-the-plot for more details.

eukaryo replied:
Thank you for your suggestion. I have modified the code to putthe legend outside of the plot, just as plotly does.

HideakiImamura commented on optuna/visualization/matplotlib/_timeline.py:
```diff
@@ -0,0 +1,119 @@
+            legend_handles.append(patches.Patch(color=color, label=state.name))
+    ax.legend(handles=legend_handles, loc="lower right")
+
+    if len(info.bars) > 0:
```
How about early returning the empty figure if `info.bars` is an empty? This is the same as other visualization modules like https://github.com/optuna/optuna/blob/master/optuna/visualization/matplotlib/_contour.py#L118.

eukaryo replied:
I see. I have revised the code.

HideakiImamura reviewed (commented):
Thanks for the PR. I have several comments. PTAL.

By the way, can we check the plot appearance using the visual regression test? [The CI job](https://github.com/optuna/optuna/actions/runs/4499036180/jobs/7916387750?pr=4538) does not generate the appropriate link.

eukaryo commented:
Here is an example of plot with the current modified code.
![test-fig-matplotlib-2](https://user-images.githubusercontent.com/22315555/228127248-360ce2c4-0148-4d8f-8c1e-e9a0acc2c89c.png)

toshihikoyanase reviewed (commented):
Could you update the list of plot functions in the reference, please?

https://github.com/optuna/optuna/blob/2d0093a07a7b814d34aa46a5e00b15c888c3aec0/docs/source/reference/visualization/matplotlib.rst?plain=1#L13-L21

eukaryo commented:
I have updated [optuna/docs/source/reference/visualization/matplotlib.rst](https://github.com/optuna/optuna/blob/2d0093a07a7b814d34aa46a5e00b15c888c3aec0/docs/source/reference/visualization/matplotlib.rst?plain=1#L13-L21) and fixed docstring in `plot_timeline` function,

eukaryo commented:
Oh, timeline_plot does not rendered... Maybe the reason is that https://github.com/optuna/optuna-visualization-regression-tests/pull/8 is not merged yet. I have resolved comments in that PR. I think we can merge that PR now. (although there is still some room for follow-up)

toshihikoyanase commented:
https://github.com/optuna/optuna-visualization-regression-tests/pull/8 was merged! Let me re-render the plot.

toshihikoyanase commented:
The timeline plots of Plotly backend were rendered as expected, but those of matplotlib backend weren't. I think we need https://github.com/optuna/optuna/pull/4555 to render plots using the code in PRs.

https://642648d03da4640e33ad5d1b--effervescent-dieffenbachia-ab819a.netlify.app/plot_timeline.html
<img width="1088" alt="image" src="https://user-images.githubusercontent.com/3255979/229010182-a5bcdca6-d95b-4cc6-b1d0-e2018d611940.png">

HideakiImamura reviewed (commented):
Thanks for the update. Almost LGTM. I have a minor comment. PTAL.

HideakiImamura commented:
I confirmed that the matplotlib version by using the following code.

```python
import time
import datetime

import optuna
import matplotlib.pyplot as plt


def objective(trial):
    if trial.number % 3 == 0:
        time.sleep(0.003 * trial.number)
        return 1
    elif trial.number % 3 == 1:
        time.sleep(0.002 * trial.number)
        raise optuna.TrialPruned()
    else:
        time.sleep(0.001 * trial.number)
        raise ValueError("FAILED!")


study = optuna.create_study()
study.optimize(objective, n_trials=100, catch=(ValueError,))

start = datetime.datetime.now()
complete = start - datetime.timedelta(seconds=1.0)
study.add_trial(
    optuna.trial.FrozenTrial(
        number=100,
        trial_id=-1,
        state=optuna.trial.TrialState.COMPLETE,
        value=0.0,
        values=None,
        datetime_start=start,
        datetime_complete=complete,
        params={},
        distributions={},
        user_attrs={},
        system_attrs={},
        intermediate_values={},
    )
)

plotly_fig = optuna.visualization.plot_timeline(study)
plotly_fig.write_image("hoge_plotly.png")

plt_fig = optuna.visualization.matplotlib.plot_timeline(study)
plt.savefig("hoge_plot.png")
```

The generated plots as follows.

| Plotly version | Matplotlib version |
| --- | --- |
| ![hoge_plotly](https://user-images.githubusercontent.com/38826298/230548504-77875cd1-15f5-4a7d-b549-6acfa51fb9dd.png) | ![hoge_plot](https://user-images.githubusercontent.com/38826298/230548524-b2ef1fb1-ca68-4a87-b0ad-52b2bd49f805.png) |

eukaryo commented on optuna/visualization/matplotlib/_timeline.py:
```diff
@@ -0,0 +1,119 @@
+        pass
+
+    def __call__(self, x: float, pos: Any = 0) -> str:
+        return dates.num2date(x).strftime("%H:%M:%S.%f")[:-3]
```
Agreed. I changed the code to `return dates.num2date(x).strftime("%H:%M:%S")`. 
In the previous code, argument `"%H:%M:%S.%f"` indicates to output time down to microseconds. `[:-3]` limited the output to milliseconds.

HideakiImamura reviewed (approved):
Looks very good!

toshihikoyanase commented on optuna/visualization/matplotlib/_timeline.py:
```diff
@@ -0,0 +1,119 @@
+    from optuna.visualization.matplotlib._matplotlib_imports import ticker
+
+
+class _DateFormatter_Millisecond(ticker.Formatter):
```
I don't think this class handles the millisecond information. Could you rename this class, please?

eukaryo replied:
I removed this class. The current code only displays hours, minutes, and seconds.

toshihikoyanase commented on optuna/visualization/matplotlib/_matplotlib_imports.py:
```diff
@@ -27,6 +30,9 @@
 
 __all__ = [
     "_imports",
+    "dates",
+    "ticker",
+    "patches",
```
Since these modules are only used in the timeline plot. How about removing them and use `matplotlib.dates`, `matplotlib.patches` and `matplotlib.ticker` instead?

eukaryo replied:
Agreed. I revised the two files.

toshihikoyanase commented:
Thank you for your update. Let me check again.

toshihikoyanase commented on docs/source/reference/visualization/matplotlib.rst:
```diff
@@ -18,4 +18,5 @@ optuna.visualization.matplotlib
     optuna.visualization.matplotlib.plot_param_importances
     optuna.visualization.matplotlib.plot_pareto_front
     optuna.visualization.matplotlib.plot_slice
+    optuna.visualization.matplotlib.plot_timeline
```
[Notes] The reference page was rendered as expected:
![image](https://user-images.githubusercontent.com/3255979/232654112-ce23fd0e-6776-484b-8510-329d2faa463a.png)

toshihikoyanase reviewed (approved):
Thank you for your update. Looks solid to me.

toshihikoyanase commented:
@amylase Let me merge this PR since it got two approvals.
